### PR TITLE
fix(provider): Reply with 404 whenever a manifest is received that has no leases

### PR DESCRIPTION
**Pull request by hydrogen18**
This changes the provider to send back 404 if the manifest is valid but there is no lease. I also added a test for this.

### CODE REVIEW

File: provider/manifest/manager.go [hunk 1]
```diff
@@ -187,6 +188,8 @@ loop:
 
 			if err := result.Error(); err != nil {
 				m.log.Error("error fetching data", "err", err)
+				// Fetching data failed, all requests are now in an error state
+				m.fillAllRequests(err)
```
Comment (boz) [inline provider/manifest/manager.go:192]: great catch

File: provider/manifest/manager.go [hunk 2]
```diff
@@ -259,8 +260,25 @@ func (m *manager) maybeScheduleStop() bool { // nolint:golint,unparam
...
+	}
+	m.requests = nil
+}
+
 func (m *manager) emitReceivedEvents() {
-	if m.data == nil || len(m.leases) == 0 || len(m.manifests) == 0 {
+	if len(m.leases) == 0 {
+		m.log.Debug("emit received events skips due to no leases", "data", m.data, "leases", len(m.leases), "manifests", len(m.manifests))
+		m.fillAllRequests(ErrNoLeaseForDeployment)
+		return
```
Comment (boz) [inline provider/manifest/manager.go:279]: Again good catch.  The idea here was that there are cases where the manifest might be shared before the lease is obtained, but I think it's better that we shed all of these historical ideas and simplify as much as possible.  To that end, I'd be open to slimming down this service significantly - I've never been happy with how difficult it is to follow (and I wrote it).

**Review by boz — APPROVED** [review]
Looks great.

One reason why it might be nice to store the whole request in `pendingRequests` is that we _could_ pass a context in with the request and the server loop _could_ discard requests whose context has already been cancelled.

I think it's fine as-is for now tho.

**Comment by boz** [comment]
refs #999

**Review by troian — APPROVED** [review]

**Pull request merged**